packages/flutter/lib/src/services/platform_views.dart

```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:ui';
+
+import 'package:flutter/foundation.dart';
+
+import 'system_channels.dart';
+
+/// The [PlatformViewRegistry] responsible for generating unique identifiers for platform views.
+final PlatformViewRegistry platformViewRegistry = PlatformViewRegistry._instance();
+
+/// A registry responsible for generating unique identifier for platform views.
+///
+/// A Flutter application has a single [PlatformViewRegistry] which can be accesses
+/// through the [platformViewRegistry] getter.
+///
+/// See also:
+///   * [PlatformView], a widget that shows a platform view.
+class PlatformViewRegistry {
+  PlatformViewRegistry._instance();
+
+  int _nextPlatformViewId = 0;
+
+  /// Allocates a unique identifier for a platform view.
+  ///
+  /// A platform view identifier can refer to a platform view that was never created,
+  /// a platform view that was disposed, or a platform view that is alive.
+  ///
+  /// Typically a platform view identifier is passed to a [PlatformView] widget
+  /// which creates the platform view and manages its lifecycle.
+  int getNextPlatformViewId() => _nextPlatformViewId++;
+}
+
+/// Callback signature for when a platform view was created.
+///
+/// `id` is the platform view's unique identifier.
+typedef void OnPlatformViewCreated(int id);
+
+/// Provides access to the platform views service.
+///
+/// This service allows creating and controlling Android views.
+///
+/// See also: [PlatformView].
+class PlatformViewsService {
+  PlatformViewsService._();
+
+  /// Creates a controller for a new Android view.
+  ///
+  /// `id` is an unused unique identifier generated with [platformViewsRegistry].
+  ///
+  /// `viewType` is the identifier of the Android view type to be created, a
+  /// factory for this view type must have been registered on the platform side.
+  /// Platform view factories are typically registered by plugin code.
+  /// Plugins can register a platform view factory with
+  /// [PlatformViewRegistry#registerViewFactory](/javadoc/io/flutter/plugin/platform/PlatformViewRegistry.html#registerViewFactory-java.lang.String-io.flutter.plugin.platform.PlatformViewFactory-).
+  ///
+  /// The Android view will only be created after [AndroidViewController.setSize] is called for the
+  /// first time.
+  static AndroidViewController initAndroidView({
+    @required int id,
+    @required String viewType,
+    OnPlatformViewCreated onPlatformViewCreated,
+  }) {
+    return new AndroidViewController._(
```

Comment:
Doesn't have to be done in this PR but just making sure we don't lock options out via the current design. Should we move the _create out of the AndroidViewController into this factory? What about cases where an expensive Android view already exists and want to be 'brought' into Flutter?

This would probably also indirectly mean you don't need 'waitingForSize' or have an AndroidViewController in a partially constructed state if the factory needs the size to call _create.

Reply:
We create the Android views with a custom Android Context that also depends on the viewport  size. Unfortunately this means that we cannot bring in an already existing view (as it was created with the wrong context), and we cannot create the Android view before we know the size.

I actually started with  the _create logic here (see the state before https://github.com/flutter/flutter/pull/19325/commits/67197a179792d020fe7b074695b1e78cd8c590e8 ) which also took the size, but that lead to complications in the Render object (as the responsibility of managing the view's lifecycle and sizing the view were coupled).